packages/storage/amplify_storage_s3_dart/test/storage_s3_service.dart/storage_s3_service_test.dart

```diff
@@ -482,5 +486,226 @@ void main() {
+      final testKeys = List.generate(
+        1005,
+        (index) => 'object-to-remove-${index + 1}',
+      ).toList();
+
+      setUpAll(() {
+        registerFallbackValue(
+          DeleteObjectsRequest(
+            bucket: 'fake bucket',
+            delete: Delete(objects: BuiltList(<ObjectIdentifier>[])),
+          ),
+        );
+      });
+
+      test('should invoke S3Client.deleteObjects with expected parameters',
+          () async {
+        const testOptions = S3StorageRemoveManyOptions(
+          storageAccessLevel: StorageAccessLevel.protected,
+        );
+        final testPrefix =
+            '${testOptions.storageAccessLevel.defaultPrefix}$testDelimiter';
+        final testDeleteObjectsOutput1 = DeleteObjectsOutput(
+          deleted: BuiltList(
+            testKeys
+                .take(1000 - testNumOfRemoveErrors)
+                .map((key) => DeletedObject(key: '$testPrefix$key')),
+          ),
+          errors: BuiltList(
+            testKeys
+                .skip(1000 - testNumOfRemoveErrors)
+                .take(testNumOfRemoveErrors)
+                .map(
+                  (key) => Error(
+                    key: '$testPrefix$key',
+                    message: 'some error',
+                  ),
+                ),
+          ),
+        );
+        final testDeleteObjectsOutput2 = DeleteObjectsOutput(
+          deleted: BuiltList(
+            testKeys
+                .skip(1000)
+                .take(5)
+                .map((key) => DeletedObject(key: '$testPrefix$key')),
+          ),
+        );
+
+        // response to the first 1000 delete objects request
+        when(
+          () => s3Client.deleteObjects(
+            any(that: DeleteObjectsLength(equals(1000))),
+          ),
+        ).thenAnswer((_) async => testDeleteObjectsOutput1);
+
+        // response to the remaining delete objects request
+        when(
+          () => s3Client.deleteObjects(
+            any(that: DeleteObjectsLength(equals(5))),
+          ),
+        ).thenAnswer((_) async => testDeleteObjectsOutput2);
+
+        removeManyResult = await storageS3Service.removeMany(
+          keys: testKeys,
+          options: testOptions,
+        );
+
+        final capturedRequests = verify(
+          () => s3Client.deleteObjects(captureAny<DeleteObjectsRequest>()),
+        ).captured;
+
+        // 1005 objects to remove therefore should send 2 requests for
+        // 2 batches
+        expect(capturedRequests, hasLength(2));
+
+        final capturedRequest1 = capturedRequests.first;
+        final capturedRequest2 = capturedRequests.last;
+
+        expect(capturedRequest1 is DeleteObjectsRequest, isTrue);
+        expect(capturedRequest2 is DeleteObjectsRequest, isTrue);
+
+        final request1 = capturedRequest1 as DeleteObjectsRequest;
+        final request2 = capturedRequest2 as DeleteObjectsRequest;
+
+        request1.delete.objects.asMap().forEach((index, object) async {
```

Comment:
A forEach won't await the result. This could complete after the test is done in which case the results won't be captured as part of the test.

If you use `expectLater` or an asynchronous matcher in the second arg to `expect`, it will create an asynchronous expectation which will prevent the test from finishing until all expectations complete. But with that, the first argument needs to be a future and the second argument should be an asynchronous matcher like `completes`, `completion`, `throwsA`, etc.

Reply:
Or you could precompute the awaited values before running the forEach loops.